week0004/HaiHa/cpp/PrintOutDuplicateInfo.cpp

```diff
@@ -0,0 +1,101 @@
+#include <algorithm>
+#include <fstream>
+#include <iostream>
+#include <set>
+#include <regex>
+
+class Participant
+{
+
+public:
+  Participant (std::string firstName, std::string lastName, std::string email) :
+    m_firstName(firstName),
+    m_lastName(lastName),
+    m_email(email)
+    {
+    }
+
+  bool operator< (const Participant &rhs) const
+  {
+    if (m_firstName != rhs.m_firstName)
+      return (m_firstName < rhs.m_firstName);
+    else if (m_lastName != rhs.m_lastName)
+      return (m_lastName < rhs.m_lastName);
+    else
+      return (m_email < rhs.m_email);
+  }
+
+  friend std::ostream & operator<<(std::ostream &out, Participant input)
```

Comment:
If you treat this function as a friend, you don't need to keep its implementation inside the class. Recommended idiom of larger projects:

```
class A
{
  int data;
public:
  void print(std::ostream& ost)
  {
    ost << "Data: " << data;
  }
};

std::ostream& operator<<(std::ostream& ostr, const A& a)
{
 a.print(ostr);
 return ostr;
}
```